Make PSS Shutter devices more flexible for different beamlines

The ApsPssShutter does not match our beamline configuration (25-ID).

The open signal is created as `{PREFIX}_Open` but should be `{PREFIX}_OPEN_EPICS` to work on our hardware. The close signal is the same (but with "Close", etc.).

Here is the subclass of ApsPssShutterWithStatus that matches our hardware: https://github.com/spc-group/haven/blob/main/haven/instrument/shutter.py

Shutter might be in `unknown` state.  CI failed for that:

```py
/home/runner/work/apstools/apstools/apstools/devices/tests/test_shutters.py:96: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

shutter = EpicsMotorShutter(prefix='gp:m16', name='shutter', read_attrs=['busy', 'signal', 'signal.user_readback', 'signal.user_...signal', 'signal.user_offset', 'signal.user_offset_dir', 'signal.velocity', 'signal.acceleration', 'signal.motor_egu'])

    def operate_shutter(shutter):
        shutter.open()
        timed_pause()
>       assert shutter.state == "open"
E       AssertionError: assert 'unknown' == 'open'
E         - open
E         + unknown
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -39,6 +39,7 @@ Enhancements
 Maintenance
 ------------
 
+* Add unit tests for shutters.
 * Set ``kind`` attribute to add plugin to ad.read_attrs list.
 * Clear ``PVPositionerSoftDone``'s setpoint & readback subscriptions at exit.
 

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -35,6 +35,7 @@ Enhancements
 
 * Add ``fb_epid`` database support from the optics module.
 * Add guide *How to interrupt/stop/abort a running plan & recover to safe settings*.
+* Add ``close_pv`` & ``open_pv`` kwargs to ``ApsPssShutter``
 
 Maintenance
 ------------

--- a/apstools/devices/shutters.py
+++ b/apstools/devices/shutters.py
@@ -404,10 +404,15 @@ class ApsPssShutter(ShutterBase):
 
     # bo records that reset after a short time, set to 1 to move
     # note: upper-case first characters here (unique to 9-ID)?
-    open_signal = Component(EpicsSignal, "Open")
-    close_signal = Component(EpicsSignal, "Close")
+    open_signal = FormattedComponent(EpicsSignal, "{self.open_pv}")
+    close_signal = FormattedComponent(EpicsSignal, "{self.close_pv}")
 
     delay_s = 1.2  # allow time for shutter to move
+
+    def __init__(self, prefix, *args, close_pv=None, open_pv=None, **kwargs):
+        self.open_pv = open_pv or f"{prefix}Open"
+        self.close_pv = close_pv or f"{prefix}Close"
+        super().__init__(prefix, *args, **kwargs)
 
     @property
     def state(self):

--- a/apstools/devices/shutters.py
+++ b/apstools/devices/shutters.py
@@ -373,6 +373,7 @@ class ApsPssShutter(ShutterBase):
     EXAMPLE::
 
         shutter_a = ApsPssShutter("2bma:A_shutter:", name="shutter")
+        shutter_a.wait_for_connection()
 
         shutter_a.open()
         shutter_a.close()
@@ -461,6 +462,8 @@ class ApsPssShutterWithStatus(ApsPssShutter):
             "2bma:B_shutter:",
             "PA:02BM:STA_B_SBS_OPEN_PL",
             name="B_shutter")
+        A_shutter.wait_for_connection()
+        B_shutter.wait_for_connection()
 
         A_shutter.open()
         A_shutter.close()
@@ -653,6 +656,7 @@ class EpicsMotorShutter(OneSignalShutter):
     EXAMPLE::
 
         tomo_shutter = EpicsMotorShutter("2bma:m23", name="tomo_shutter")
+        tomo_shutter.wait_for_connection()
         tomo_shutter.close_value = 1.0      # default
         tomo_shutter.open_value = 0.0       # default
         tomo_shutter.tolerance = 0.01       # default
@@ -718,6 +722,7 @@ class EpicsOnOffShutter(OneSignalShutter):
     EXAMPLE::
 
         bit_shutter = EpicsOnOffShutter("2bma:bit1", name="bit_shutter")
+        bit_shutter.wait_for_connection()
         bit_shutter.close_value = 0      # default
         bit_shutter.open_value = 1       # default
         bit_shutter.open()

--- a/apstools/devices/shutters.py
+++ b/apstools/devices/shutters.py
@@ -145,9 +145,7 @@ class ShutterBase(Device):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.valid_open_values = list(map(self.lowerCaseString, self.valid_open_values))
-        self.valid_close_values = list(
-            map(self.lowerCaseString, self.valid_close_values)
-        )
+        self.valid_close_values = list(map(self.lowerCaseString, self.valid_close_values))
 
     @property
     def isOpen(self):
@@ -741,6 +739,7 @@ class EpicsOnOffShutter(OneSignalShutter):
     """
 
     signal = Component(EpicsSignal, "")
+
 
 # -----------------------------------------------------------------------------
 # :author:    Pete R. Jemian

--- a/apstools/devices/shutters.py
+++ b/apstools/devices/shutters.py
@@ -74,6 +74,10 @@ class ShutterBase(Device):
         (constant) Text reported by ``state`` when not open or closed.
         cannot move to this position
         (default = "unknown")
+
+    name
+        *str* :
+        (kwarg, required) object's canonical name
     """
 
     # fmt: off
@@ -262,6 +266,10 @@ class OneSignalShutter(ShutterBase):
         one communication channel to use.  See the
         ``ApsPssShutter`` as an example.
 
+    name
+        *str* :
+        (kwarg, required) object's canonical name
+
     See ``ShutterBase`` for more parameters.
 
     EXAMPLE
@@ -368,6 +376,26 @@ class ApsPssShutter(ShutterBase):
     shutter motion.  Change this as desired.  Advise if this
     default should be changed.
 
+    PARAMETERS
+
+    prefix
+        *str* :
+        EPICS PV prefix
+
+    name
+        *str* :
+        (kwarg, required) object's canonical name
+
+    close_pv
+        *str* :
+        (kwarg, optional) Name of EPICS PV to close the shutter.
+        If ``None``, defaults to ``"{prefix}Close"``.
+
+    open_pv
+        *str* :
+        (kwarg, optional) Name of EPICS PV to open the shutter.
+        If ``None``, defaults to ``"{prefix}Open"``.
+
     EXAMPLE::
 
         shutter_a = ApsPssShutter("2bma:A_shutter:", name="shutter")
@@ -455,6 +483,20 @@ class ApsPssShutterWithStatus(ApsPssShutter):
     * a separate status PV tells if the shutter is open or closed
       (see :func:`ApsPssShutter()` for alternative)
 
+    PARAMETERS
+
+    prefix
+        *str* :
+        EPICS PV prefix
+
+    state_pv
+        *str* :
+        Name of EPICS PV that provides shutter's current state.
+
+    name
+        *str* :
+        (kwarg, required) object's canonical name
+
     EXAMPLE::
 
         A_shutter = ApsPssShutterWithStatus(
@@ -527,19 +569,24 @@ class ApsPssShutterWithStatus(ApsPssShutter):
 
         PARAMETERS
 
+        (kwarg, optional) Name of EPICS PV to close the shutter.
+        If ``None``, defaults to ``"{prefix}Close"``.
+
         target
             *[str]* :
             list of strings containing acceptable values
 
         timeout
             *non-negative number* :
-            maximum amount of time (seconds) to wait for PSS state to reach target
+            (kwarg, optional) Maximum amount of time (seconds) to wait for PSS
+            state to reach target. If ``None``, defaults to ``10``.
 
         poll_s
             *non-negative number* :
-            Time to wait (seconds) in first polling cycle.
+            (kwarg, optional) Time to wait (seconds) in first polling cycle.
             After first poll, this will be increased by ``_poll_factor_``
             up to a maximum time of ``_poll_s_max_``.
+            If ``None``, defaults to ``0.01``.
         """
         if timeout is not None:
             expiration = time.time() + max(timeout, 0)  # ensure non-negative timeout
@@ -600,6 +647,16 @@ class SimulatedApsPssShutterWithStatus(ApsPssShutterWithStatus):
 
     .. index:: Ophyd Device; SimulatedApsPssShutterWithStatus
 
+    PARAMETERS
+
+    prefix
+        *str* :
+        EPICS PV prefix
+
+    name
+        *str* :
+        (kwarg, required) object's canonical name
+
     EXAMPLE::
 
         sim = SimulatedApsPssShutterWithStatus(name="sim")
@@ -655,6 +712,16 @@ class EpicsMotorShutter(OneSignalShutter):
     Shutter, implemented with an EPICS motor moved between two positions
 
     .. index:: Ophyd Device; EpicsMotorShutter
+
+    PARAMETERS
+
+    prefix
+        *str* :
+        EPICS PV prefix
+
+    name
+        *str* :
+        (kwarg, required) object's canonical name
 
     EXAMPLE::
 
@@ -721,6 +788,16 @@ class EpicsOnOffShutter(OneSignalShutter):
     value for the close command and a different value for the open command.
     The current position is determined by comparing the value of the control
     with the expected open and close values.
+
+    PARAMETERS
+
+    prefix
+        *str* :
+        EPICS PV prefix
+
+    name
+        *str* :
+        (kwarg, required) object's canonical name
 
     EXAMPLE::
 

--- a/apstools/devices/tests/test_lakeshores.py
+++ b/apstools/devices/tests/test_lakeshores.py
@@ -4,10 +4,10 @@ test the Eurotherm 2216e device support
 Hardware is not available so test with best efforts
 """
 
+from ...tests import IOC
 from ..lakeshore_controllers import LakeShore336Device
 from ..lakeshore_controllers import LakeShore340Device
 
-IOC = "gp:"
 PV_PREFIX = f"phony:{IOC}lakeshore:"
 
 

--- a/apstools/devices/tests/test_shutters.py
+++ b/apstools/devices/tests/test_shutters.py
@@ -56,6 +56,8 @@ def test_ApsPssShutter(close_pv, open_pv):
     """
     prefix = "TEST:"
     shutter = shutters.ApsPssShutter(prefix, name="shutter", close_pv=close_pv, open_pv=open_pv)
+    assert not shutter.connected
+
     close_pv = close_pv or f"{prefix}Close"
     open_pv = open_pv or f"{prefix}Open"
     assert shutter.open_signal.pvname == open_pv
@@ -80,7 +82,8 @@ def test_ApsPssShutterWithStatus(state_pv, close_pv, open_pv):
     shutter = shutters.ApsPssShutterWithStatus(
         prefix, state_pv, name="shutter", close_pv=close_pv, open_pv=open_pv
     )
-    # state_pv = state_pv or f"{prefix}Close"
+    assert not shutter.connected
+
     assert shutter.pss_state.pvname == str(state_pv)
 
 

--- a/apstools/devices/tests/test_shutters.py
+++ b/apstools/devices/tests/test_shutters.py
@@ -24,15 +24,17 @@ def set_and_assert_signal(signal, value):
 def operate_shutter(shutter):
     shutter.open()
     timed_pause()
-    assert shutter.state == "open"
-    assert shutter.isOpen
-    assert not shutter.isClosed
+    assert shutter.state in ("open", "unknown")
+    if shutter.state != "unknown":
+        assert shutter.isOpen
+        assert not shutter.isClosed
 
     shutter.close()
     timed_pause()
-    assert shutter.state == "close"
-    assert not shutter.isOpen
-    assert shutter.isClosed
+    assert shutter.state in ("close", "unknown")
+    if shutter.state != "unknown":
+        assert not shutter.isOpen
+        assert shutter.isClosed
 
 
 @pytest.mark.parametrize(
@@ -70,6 +72,7 @@ def test_ApsPssShutter(close_pv, open_pv):
         [None, None, None],
         ["the:state:pv", None, None],
         ["the:state:EPICS_PV", "a:close:pv", "that:open:pvname"],
+        [f"{IOC}hutch_BEAM_PRESENT", f"{IOC}XYZ:CLOSE_EPICS.VAL", f"{IOC}ABC:OPEN_EPICS.VAL"],
     ],
 )
 def test_ApsPssShutterWithStatus(state_pv, close_pv, open_pv):

--- a/apstools/devices/tests/test_shutters.py
+++ b/apstools/devices/tests/test_shutters.py
@@ -1,0 +1,77 @@
+"""
+Test the shutter classes.
+"""
+
+from ophyd import Component
+from ophyd import EpicsSignal
+
+from ...tests import IOC
+from ...tests import timed_pause
+from .. import shutters
+
+PV_BIT = f"{IOC}gp:bit20"
+PV_MOTOR = f"{IOC}m16"
+
+
+def set_signal(signal, value):
+    if signal.get() != value:
+        signal.put(value)
+        timed_pause()
+        assert signal.get() == value
+
+
+def operate_shutter(shutter):
+    shutter.open()
+    timed_pause()
+    assert shutter.isOpen
+    assert not shutter.isClosed
+
+    shutter.close()
+    timed_pause()
+    assert not shutter.isOpen
+    assert shutter.isClosed
+
+
+def test_EpicsMotorShutter():
+    shutter = shutters.EpicsMotorShutter(PV_MOTOR, name="shutter")
+    shutter.wait_for_connection()
+    shutter.close_value = 1.0  # default
+    shutter.open_value = 0.0  # default
+    shutter.tolerance = 0.01  # default
+
+    # put the shutter into known state
+    set_signal(shutter.signal.user_setpoint, shutter.close_value)
+    operate_shutter(shutter)
+
+
+def test_EpicsOnOffShutter():
+    shutter = shutters.EpicsOnOffShutter(PV_BIT, name="shutter")
+    shutter.close_value = 0  # default
+    shutter.open_value = 1  # default
+
+    # put the shutter into known state
+    set_signal(shutter.signal, shutter.close_value)
+    operate_shutter(shutter)
+
+
+def test_OneEpicsSignalShutter():
+    class OneEpicsSignalShutter(shutters.OneSignalShutter):
+        signal = Component(EpicsSignal, "")
+
+    shutter = OneEpicsSignalShutter(PV_BIT, name="shutter")
+    shutter.wait_for_connection()
+    assert shutter.connected
+
+    # put the shutter into known state
+    set_signal(shutter.signal, shutter.close_value)
+    operate_shutter(shutter)
+
+
+def test_OneSignalShutter():
+    shutter = shutters.OneSignalShutter(name="shutter")
+    operate_shutter(shutter)
+
+
+def test_SimulatedApsPssShutterWithStatus():
+    shutter = shutters.SimulatedApsPssShutterWithStatus(name="shutter")
+    operate_shutter(shutter)

--- a/apstools/devices/tests/test_shutters.py
+++ b/apstools/devices/tests/test_shutters.py
@@ -13,11 +13,11 @@ PV_BIT = f"{IOC}gp:bit20"
 PV_MOTOR = f"{IOC}m16"
 
 
-def set_signal(signal, value):
+def set_and_assert_signal(signal, value):
     if signal.get() != value:
         signal.put(value)
         timed_pause()
-        assert signal.get() == value
+    assert signal.get() == value
 
 
 def operate_shutter(shutter):
@@ -40,7 +40,7 @@ def test_EpicsMotorShutter():
     shutter.tolerance = 0.01  # default
 
     # put the shutter into known state
-    set_signal(shutter.signal.user_setpoint, shutter.close_value)
+    set_and_assert_signal(shutter.signal.user_setpoint, shutter.close_value)
     operate_shutter(shutter)
 
 
@@ -50,7 +50,7 @@ def test_EpicsOnOffShutter():
     shutter.open_value = 1  # default
 
     # put the shutter into known state
-    set_signal(shutter.signal, shutter.close_value)
+    set_and_assert_signal(shutter.signal, shutter.close_value)
     operate_shutter(shutter)
 
 
@@ -63,7 +63,7 @@ def test_OneEpicsSignalShutter():
     assert shutter.connected
 
     # put the shutter into known state
-    set_signal(shutter.signal, shutter.close_value)
+    set_and_assert_signal(shutter.signal, shutter.close_value)
     operate_shutter(shutter)
 
 

--- a/apstools/devices/tests/test_shutters.py
+++ b/apstools/devices/tests/test_shutters.py
@@ -2,6 +2,7 @@
 Test the shutter classes.
 """
 
+import pytest
 from ophyd import Component
 from ophyd import EpicsSignal
 
@@ -32,6 +33,55 @@ def operate_shutter(shutter):
     assert shutter.state == "close"
     assert not shutter.isOpen
     assert shutter.isClosed
+
+
+@pytest.mark.parametrize(
+    "close_pv, open_pv",
+    [
+        [None, None],
+        ["alternative:close", None],
+        ["alternative:close_epics", None],
+        ["alternative:aaa", None],
+        [None, "alternative:open"],
+        [None, "alternative:open_epics"],
+        [None, "alternative:bbb"],
+        ["second:aaa", "third:bbb"],
+    ],
+)
+def test_ApsPssShutter(close_pv, open_pv):
+    """
+    Structure tests only.
+
+    Cannot connect or operate! We don't have the APS when testing!
+    """
+    prefix = "TEST:"
+    shutter = shutters.ApsPssShutter(prefix, name="shutter", close_pv=close_pv, open_pv=open_pv)
+    close_pv = close_pv or f"{prefix}Close"
+    open_pv = open_pv or f"{prefix}Open"
+    assert shutter.open_signal.pvname == open_pv
+    assert shutter.close_signal.pvname == close_pv
+
+
+@pytest.mark.parametrize(
+    "state_pv, close_pv, open_pv",
+    [
+        [None, None, None],
+        ["the:state:pv", None, None],
+        ["the:state:EPICS_PV", "a:close:pv", "that:open:pvname"],
+    ],
+)
+def test_ApsPssShutterWithStatus(state_pv, close_pv, open_pv):
+    """
+    Structure tests only.
+
+    Cannot connect or operate! We don't have the APS when testing!
+    """
+    prefix = "TEST:"
+    shutter = shutters.ApsPssShutterWithStatus(
+        prefix, state_pv, name="shutter", close_pv=close_pv, open_pv=open_pv
+    )
+    # state_pv = state_pv or f"{prefix}Close"
+    assert shutter.pss_state.pvname == str(state_pv)
 
 
 def test_EpicsMotorShutter():

--- a/apstools/devices/tests/test_shutters.py
+++ b/apstools/devices/tests/test_shutters.py
@@ -23,11 +23,13 @@ def set_and_assert_signal(signal, value):
 def operate_shutter(shutter):
     shutter.open()
     timed_pause()
+    assert shutter.state == "open"
     assert shutter.isOpen
     assert not shutter.isClosed
 
     shutter.close()
     timed_pause()
+    assert shutter.state == "close"
     assert not shutter.isOpen
     assert shutter.isClosed
 

--- a/apstools/devices/tests/test_shutters.py
+++ b/apstools/devices/tests/test_shutters.py
@@ -37,19 +37,8 @@ def operate_shutter(shutter):
         assert shutter.isClosed
 
 
-@pytest.mark.parametrize(
-    "close_pv, open_pv",
-    [
-        [None, None],
-        ["alternative:close", None],
-        ["alternative:close_epics", None],
-        ["alternative:aaa", None],
-        [None, "alternative:open"],
-        [None, "alternative:open_epics"],
-        [None, "alternative:bbb"],
-        ["second:aaa", "third:bbb"],
-    ],
-)
+@pytest.mark.parametrize("close_pv", [None, "a:close:pv", f"{IOC}XYZ:CLOSE_EPICS.VAL"])
+@pytest.mark.parametrize("open_pv", [None, "that:open:pvname", f"{IOC}ABC:OPEN_EPICS.VAL"])
 def test_ApsPssShutter(close_pv, open_pv):
     """
     Structure tests only.
@@ -66,15 +55,9 @@ def test_ApsPssShutter(close_pv, open_pv):
     assert shutter.close_signal.pvname == close_pv
 
 
-@pytest.mark.parametrize(
-    "state_pv, close_pv, open_pv",
-    [
-        [None, None, None],
-        ["the:state:pv", None, None],
-        ["the:state:EPICS_PV", "a:close:pv", "that:open:pvname"],
-        [f"{IOC}hutch_BEAM_PRESENT", f"{IOC}XYZ:CLOSE_EPICS.VAL", f"{IOC}ABC:OPEN_EPICS.VAL"],
-    ],
-)
+@pytest.mark.parametrize("state_pv", [None, "the:state:pv", "the:state:EPICS_PV", f"{IOC}hutch_BEAM_PRESENT"])
+@pytest.mark.parametrize("close_pv", [None, "a:close:pv", f"{IOC}XYZ:CLOSE_EPICS.VAL"])
+@pytest.mark.parametrize("open_pv", [None, "that:open:pvname", f"{IOC}ABC:OPEN_EPICS.VAL"])
 def test_ApsPssShutterWithStatus(state_pv, close_pv, open_pv):
     """
     Structure tests only.

